File: day12/Main.hs
import           Text.Parsec             hiding ( Empty )
import           Text.Parsec.String             ( Parser )

data Command
  = North Int
  | South Int
  | East Int
  | West Int
  | RotL Int
  | RotR Int
  | Fwd Int

type Position = (Int, Int)

parseCommand :: Parser Command
parseCommand =
  let parseUnsigned :: Parser Int
      parseUnsigned = read <$> many1 digit
  in  (string "N" >> North <$> parseUnsigned)
        <|> (string "S" >> South <$> parseUnsigned)
        <|> (string "E" >> East <$> parseUnsigned)
        <|> (string "W" >> West <$> parseUnsigned)
        <|> (string "L" >> RotL . flip div 90 <$> parseUnsigned)
        <|> (string "R" >> RotR . flip div 90 <$> parseUnsigned)
        <|> (string "F" >> Fwd <$> parseUnsigned)

rotl :: Int -> Position -> Position
-- (1, 2) -> (-2, 1) -> (-1, -2) -> (2, -1) -> (1, 2) -> ...
rotl 0 p        = p
rotl n p@(x, y) = if n >= 4 then rotl (n `mod` 4) p else rotl (n - 1) (-y, x)

rotr :: Int -> Position -> Position
rotr n = rotl (4 - (n `mod` 4))

route1 :: [Command] -> Position
route1 cmds =
  let step :: (Position, Position) -> Command -> (Position, Position)
      step ((x, y), dir         ) (North by) = ((x, y + by), dir)
      step ((x, y), dir         ) (South by) = ((x, y - by), dir)
      step ((x, y), dir         ) (East  by) = ((x + by, y), dir)
      step ((x, y), dir         ) (West  by) = ((x - by, y), dir)
      step (p     , dir         ) (RotL  by) = (p, rotl by dir)
      step (p     , dir         ) (RotR  by) = (p, rotr by dir)
      step ((x, y), dir@(dx, dy)) (Fwd   by) = ((x + by * dx, y + by * dy), dir)
  in  fst $ foldl step ((0, 0), (1, 0)) cmds

route2 :: [Command] -> Position
route2 cmds =
  let step :: (Position, Position) -> Command -> (Position, Position)
      step (p     , (wx, wy)   ) (North by) = (p, (wx, wy + by))
      step (p     , (wx, wy)   ) (South by) = (p, (wx, wy - by))
      step (p     , (wx, wy)   ) (East  by) = (p, (wx + by, wy))
      step (p     , (wx, wy)   ) (West  by) = (p, (wx - by, wy))
      step (p     , wp         ) (RotL  by) = (p, rotl by wp)
      step (p     , wp         ) (RotR  by) = (p, rotr by wp)
      step ((x, y), wp@(wx, wy)) (Fwd   by) = ((x + by * wx, y + by * wy), wp)
  in  fst $ foldl step ((0, 0), (10, 1)) cmds

l1 :: Position -> Int
l1 (x, y) = abs x + abs y

main = do
  input <- lines <$> readFile "input.txt"
  let Right cmds = mapM (parse parseCommand "") input
  print "Part 1:"
  print $ l1 $ route1 cmds
  print "Part 2:"
  print $ l1 $ route2 cmds

File: day12/main.ml
(*** UTILS ***)
let read_file filename =
  let lines = ref [] in
  let channel = open_in filename in
  try
    while true do
      lines := input_line channel :: !lines
    done;
    !lines
  with End_of_file ->
    close_in channel;
    List.rev !lines

let println a =
  Format.print_string (a ^ "\n");
  Format.print_flush ()

exception Exit of int

let error a =
  println a;
  raise (Exit 1)

(*** AoC ***)

type command =
  | North of int
  | South of int
  | East of int
  | West of int
  | Rotl of int
  | Rotr of int
  | Fwd of int

let parse_command line =
  let by = int_of_string (String.sub line 1 (String.length line - 1)) in
  match line.[0] with
  | 'N' -> North by
  | 'S' -> South by
  | 'E' -> East by
  | 'W' -> West by
  | 'L' -> Rotl (by / 90)
  | 'R' -> Rotr (by / 90)
  | 'F' -> Fwd by
  | _ -> error "Bad state: not a known command"

let rec rotl by pos =
  match (by, pos) with
  (* (1, 2) -> (-2, 1) -> (-1, -2) -> (2, -1) -> (1, 2) -> ... *)
  | 0, p -> p
  | n, (x, y) when n < 4 -> rotl (n - 1) (-y, x)
  | n, p -> rotl (n mod 4) p

let rotr by pos = rotl (4 - (by mod 4)) pos

let route1 cmds =
  let rec step cmds pos dir =
    match (cmds, pos, dir) with
    | [], pos, _ -> pos
    | North by :: rest, (x, y), d -> step rest (x, y + by) d
    | South by :: rest, (x, y), d -> step rest (x, y - by) d
    | East by :: rest, (x, y), d -> step rest (x + by, y) d
    | West by :: rest, (x, y), d -> step rest (x - by, y) d
    | Rotl by :: rest, p, d -> step rest p (rotl by d)
    | Rotr by :: rest, p, d -> step rest p (rotr by d)
    | Fwd by :: rest, (x, y), ((dx, dy) as d) ->
        step rest (x + (by * dx), y + (by * dy)) d
  in
  step cmds (0, 0) (1, 0)

let route2 cmds =
  let rec step cmds pos dir =
    match (cmds, pos, dir) with
    | [], pos, _ -> pos
    | North by :: rest, p, (wx, wy) -> step rest p (wx, wy + by)
    | South by :: rest, p, (wx, wy) -> step rest p (wx, wy - by)
    | East by :: rest, p, (wx, wy) -> step rest p (wx + by, wy)
    | West by :: rest, p, (wx, wy) -> step rest p (wx - by, wy)
    | Rotl by :: rest, p, wp -> step rest p (rotl by wp)
    | Rotr by :: rest, p, wp -> step rest p (rotr by wp)
    | Fwd by :: rest, (x, y), ((wx, wy) as wp) ->
        step rest (x + (by * wx), y + (by * wy)) wp
  in
  step cmds (0, 0) (10, 1)

let l1 (x, y) = abs x + abs y

let main () =
  let lines = read_file "input.txt" in
  let commands = List.map parse_command lines in

  println "Part 1:";
  println (string_of_int (l1 (route1 commands)));

  println "Part 2:";
  println (string_of_int (l1 (route2 commands)))

let () = main ()
